src/python/pants/engine/internals/platform_rules.py

```diff
@@ -15,5 +24,43 @@ def current_platform(env_tgt: EnvironmentTarget) -> Platform:
     return Platform(env_tgt.val[DockerPlatformField].normalized_value)
 
 
+@rule
+async def complete_environment_vars(
+    session_values: SessionValues, env_tgt: EnvironmentTarget
+) -> CompleteEnvironmentVars:
+    if not env_tgt.val or not env_tgt.val.has_field(DockerImageField):
+        return session_values[CompleteEnvironmentVars]
+    env_process_result = await Get(
+        ProcessResult,
+        Process(
+            ["env"],
```

Comment:
Unfortunately it looks like `env` is dumb when it comes to multi-line environment variables (there is no escaping around the key).

It _looks_ like `BashBinary` doesn't actually depend on the environment (which is ... unexpected but useful), and so this could probably be a bash loop over the output of `env` which took only keys, and then rendered the key/value pairs as quoted...?

Reply:
`env -0`? But maybe not all env support this. We really ought to use our own static binary injected in when we get time.

Reply:
https://github.com/pantsbuild/pants/pull/16876#discussion_r972218152 some further discussion.

Reply:
> `env -0`? But maybe not all env support this.

macOS doesn't... but that isn't actually relevant here probably.

@Eric-Arellano : Maybe see whether you can get this working with `env -0` (and then splitting the lines on null bytes rather than newlines: `s.split('\0')`)?

While I do think that native binaries for this are in our future, I'm really not eager to tackle that.

Reply:
I'm suggesting we just build the binaries once and check them in as binary resources. I've already written the code for the binary; it is just a matter of building them and checking those bianries into the repository.

Reply:
And I wrote it in Rust:

```rust
fn main() {
  for (k, v) in std::env::vars() {
    // Note that the delimiter `,` is not valid in an env var name.
    // We use `\0` rather than `\n` to handle newlines in env var values.
    print!("{k},{v}\0");
  }
}
```

Agreed with Tom on checking in the built binaries. Almost done with this.

Reply:
Building the Rust version in release mode results in a 443,655 byte binary on macOS. The C version is way more efficient and tractable for checking in to the repository.

Reply:
> I'm suggesting we just build the binaries once and check them in as binary resources. I've already written the code for the binary; it is just a matter of building them and checking those bianries into the repository.

"Just" building a new binary and checking it in is not a trivial endeavour. You need to build it for at least two os/platforms in a sufficiently old OS that libc compatibility won't be an issue (like Centos5 for example), and even then it won't actually work for Windows as John pointed out. You will also need an explanation of how to build new copies.

If the end result is that you have built `env -0`, which is all we actually need in this case, that doesn't seem worthwhile.

Reply:
Then we should just require that images have `env` or equivalent available.

Reply:
> Then we should just require that images have env or equivalent available.

Agreed. And if this is blocking users, then we can implement https://github.com/pantsbuild/pants/issues/16887. Although, I'm not sure that ticket is even worth having -- we also require a bunch of other binaries like `bash` and `unzip`. It seems reasonable to likewise require `env` if we need those.

For Windows, perhaps we handle that in its own way when the time comes?

Reply:
> For Windows, perhaps we handle that in its own way when the time comes?

Yea. Can switch on the `Platform` to use a different strategy if needed.